File: tests/test_fru_cache.cpp
#include <iostream>
#include <cassert>
#include "../src/cpp/fru_cache.cpp"

void test_basic_set_get() {
    FRUCache<std::string, int> cache(2, 2);
    cache.set("a", 1);
    auto val = cache.get("a");
    assert(val.has_value() && val.value() == 1);
    std::cout << "✔ test_basic_set_get passed\n";
}

void test_eviction_behavior() {
    FRUCache<std::string, int> cache(2, 2);
    cache.set("a", 1);
    cache.set("b", 2);
    cache.set("c", 3);  // should evict one of a or b
    auto val = cache.get("a");
    std::cout << "✔ test_eviction_behavior passed (a was " << (val.has_value() ? "retained" : "evicted") << ")\n";
}

void test_frequency_bias() {
    FRUCache<std::string, int> cache(2, 2);
    for (int i = 0; i < 5; ++i) {
        cache.set("x", 10);
        cache.get("x");
    }
    cache.set("y", 20);
    cache.set("z", 30);
    auto val = cache.get("x");
    assert(val.has_value() && val.value() == 10);
    std::cout << "✔ test_frequency_bias passed\n";
}

int main() {
    test_basic_set_get();
    test_eviction_behavior();
    test_frequency_bias();
    std::cout << "All FRUCache C++ tests passed.\n";
    return 0;
}

File: tests/test_fru_cache.py
import unittest
from fru_cache import FRUCache

class TestFRUCache(unittest.TestCase):
    def test_basic_set_get(self):
        cache = FRUCache(active_size=2, ghost_size=2)
        cache.set("a", 1)
        self.assertEqual(cache.get("a"), 1)

    def test_eviction(self):
        cache = FRUCache(active_size=2, ghost_size=2)
        cache.set("a", 1)
        cache.set("b", 2)
        cache.set("c", 3)
        result = cache.get("a")
        self.assertIn(result, [None, 1])  # May be evicted or promoted depending on timing

    def test_frequency_bias(self):
        cache = FRUCache(active_size=2, ghost_size=2)
        for _ in range(5):
            cache.set("x", 100)
            cache.get("x")
        cache.set("y", 200)
        cache.set("z", 300)
        self.assertEqual(cache.get("x"), 100)  # Should survive due to frequent access

if __name__ == "__main__":
    unittest.main()